Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp

```diff
@@ -50,29 +50,47 @@ ApplicationManifest ApplicationManifestParser::parse(const String& source, const
     return parser.parseManifest(source, manifestURL, documentURL);
 }
 
+std::optional<ApplicationManifest> ApplicationManifestParser::parseWithValidation(const String& source, const URL& manifestURL, const URL& documentURL)
+{
+    if (!manifestURL.isValid() || !documentURL.isValid())
+        return std::nullopt;
+
+    ApplicationManifestParser parser { nullptr };
+    auto manifest = parser.parseManifest(source, manifestURL, documentURL, true);
+    if (manifest.manifestURL.isNull())
+        return std::nullopt;
+
+    return manifest;
+}
+
 ApplicationManifestParser::ApplicationManifestParser(RefPtr<Document> document)
     : m_document(document)
 {
 }
 
-ApplicationManifest ApplicationManifestParser::parseManifest(const String& text, const URL& manifestURL, const URL& documentURL)
+ApplicationManifest ApplicationManifestParser::parseManifest(const String& text, const URL& manifestURL, const URL& documentURL, bool shouldFinishOnError)
```

Comment:
This `shouldFinishOnError` boolean parameter isn't nice. Also, ideally, we wouldn't construct an `ApplicationManifest` at all when JSON parsing fails. My suggestion would be:
1. Move the logic to convert the String to a JSObject (e.g. do the JSON parsing) to a new utility function and call it at the call sites of parseManifest()
2. Rename parseManifest() to constructManifest() and have it take the parsed JSON result as input
3. Your new parseWithValidation() function only needs to call the parsing function but never attempt to construct an ApplicationManifest.

Reply:
Implemented the suggestion in the new patch, except renaming `parseManifest` since all functions in the class that takes `JSON::Object& manifest` are using `parse` as prefix.
But let me know if you still prefer `constructManifest ` and I will change it.